src/TraceEvent/EventPipe/EventCache.cs

```diff
@@ -0,0 +1,282 @@
+        }
+
+        private unsafe void SortAndDispatch(long stopTimestamp)
+        {
+            // This sort could be made faster by using a min-heap but this is a simple place to start
+            List<Queue<EventMarker>> threadQueues = new List<Queue<EventMarker>>(_threads.Values.Select(t => t.Events));
+            while(true)
+            {
+                long lowestTimestamp = stopTimestamp;
+                Queue<EventMarker> oldestEventQueue = null;
+                foreach(Queue<EventMarker> threadQueue in threadQueues)
+                {
+                    if(threadQueue.Count == 0)
+                    {
+                        continue;
+                    }
+                    long eventTimestamp = threadQueue.Peek().Header.TimeStamp;
+                    if (eventTimestamp < lowestTimestamp)
+                    {
+                        oldestEventQueue = threadQueue;
+                        lowestTimestamp = eventTimestamp;
+                    }
+                }
+                if(oldestEventQueue == null)
+                {
+                    break;
+                }
+                else
+                {
+                    EventMarker eventMarker = oldestEventQueue.Dequeue();
+                    OnEvent?.Invoke(ref eventMarker.Header);
+                }
+            }
+
+            // If the app creates and destroys threads over time we need to flush old threads
+            // from the cache or memory usage will grow unbounded. AddThread handles the
+            // the thread objects but the storage for the queue elements also does not shrink
+            // below the high water mark unless we free it explicitly. Although not unbounded
+            // growth our current runtime policy reads ahead up to 10,000 events ahead per-thread
+            // * 5000 threads * 24 bytes = ~1GB. It would be nice not to leave that much memory
+            // laying around probably mostly unused.
+            foreach(Queue<EventMarker> q in threadQueues)
+            {
+                if(q.Count == 0)
+                {
+                    q.TrimExcess();
+                }
+            }
+        }
+
+        private void AddThread(long captureThreadId, EventCacheThread thread)
+        {
+            // To ensure we don't have unbounded growth we evict old threads to make room
+            // for new ones. Evicted threads can always be re-added later if they log again
+            // but there are two consequences:
+            // a) We won't detect lost events on that thread after eviction
+            // b) If the thread still had events pending dispatch they will be lost
+            // We pick the thread that has gone the longest since it last logged an event
+            // under the presumption that it is probably dead, has no events, and won't
+            // log again.
+            //
+            // In the future if we had explicit thread death notification events we could keep
```

Comment:
I believe there is an event for this, but probably not good to depend upon it unless we make it explicitly required for operation of EventPipe.

Reply:
Agreed. I wasn't planning to take a dependency on any runtime events at this layer, I want to treat them as opaque. For the thread death notification we'd probably want some properties that don't come for free with generic runtime events:

a) Guaranteed emission in the stream, regardless of memory pressure on the buffer pool
b) No events come after the notification on the same thread
c) Every thread which has ever logged an event to the stream sends the notification (regardless of whether it ran managed code or has an entry in the runtime ThreadStore).

My assumption if we did it is that it would be a dedicated top-level object in the stream similar to SequencePointBlock and it would have dedicated code in EventPipe that emits it.

Reply:
Yup, makes sense.